Layerscape LS1028ARDB/LS1046AFRWY/LX2160ARDB support

yangbolu1991 opened the pull request:
The pull request patches are from NXP Layerscape OpenWrt 20.09 release. (Reworked and cleaned up)
link: https://source.codeaurora.org/external/qoriq/qoriq-components/openwrt/
tag: layerscape-20.09

The pull request cherry-picked only BSP level changes of layerscape-20.09.
1. For layerscape user space packages patches, will upstream later to package.git.
2. Some sysupgrade fix-up not cherry-picked, since below pull request also included the fix-up.
But either of the two pull request has to rebase once one is merge.
https://github.com/openwrt/openwrt/pull/3345

This pull request is mainly to update layerscape components to latest LSDK release (LSDK-20.04-update-290520),
and to add three new boards LS1028ARDB/LS1046AFRWY/LX2160ARDB.

Please help to review and merge.
Thanks a lot.

yangbolu1991 committed:
layerscape: package crypto-hw-caam driver

This driver is to support Freescale Cryptographic Accelerator
and Assurance Module (CAAM).

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
layerscape: add LS1028ARDB board support

The LS1028A reference design board (RDB) is a computing, evaluation,
and development platform that supports industrial IoT applications,
human machine interface solutions, and industrial networking.

The LS1028A processor is dual-core 64-bit Arm Cortex-A72.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

biwenli committed:
kernel: package NXP PCF85263/85363 RTC driver

This patch packages NXP PCF85263/85363 RTC driver.

Signed-off-by: Biwen Li <biwen.li@nxp.com>
Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

biwenli committed:
layerscape: fix version number of restool

The restool Makefile obtains version number from its
git info, while OpenWrt builds just restool source code
without git info. So fix the version number in
restool Makefile in Build/Compile by using PKG_VERSION directly.

Signed-off-by: Biwen Li <biwen.li@nxp.com>
Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
layerscape: add workaround for restool issue

Add some delays in restool script as workaround to resolve below issue.

root@OpenWrt:/# ls-listni
open() failed : Invalid argument
open() failed : Invalid argument

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
layerscape: some clean up for config-5.4

Some clean up for config-5.4 to remove configs not needed,
and enable configs supported.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
kernel: add missing symbol CONFIG_AHCI_XGENE

Add missing symbol CONFIG_AHCI_XGENE.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

adschm commented on package/firmware/layerscape/ls-dpl/Makefile:
```diff
@@ -40,6 +40,10 @@ define Build/InstallDev
+	$(CP) $(PKG_BUILD_DIR)/config/lx2160a/RDB/dpl-eth.19.dtb \
+		$(STAGING_DIR_IMAGE)/fsl_lx2160a-rdb-dpl.dtb
+	$(CP) $(PKG_BUILD_DIR)/config/lx2160a/RDB/dpc-usxgmii.dtb \
+		$(STAGING_DIR_IMAGE)/fsl_lx2160a-rdb-dpc.dtb
```
Please bump PKG_RELEASE in cases where you add stuff without changing upstream version.

yangbolu1991 replied:
Ok.

adschm commented on target/linux/layerscape/base-files/etc/board.d/02_network:
```diff
@@ -14,6 +14,19 @@ case "$(board_name)" in
 		ucidef_set_interface_lan "eth0 eth1 eth2 eth3 eth6"
 		ucidef_set_interface_wan "eth4"
 		;;
+	fsl,ls1028a-rdb*)
```
Please don't use wildcards for board names in base-files. This makes it impossible to grep for a specific device string.

yangbolu1991 replied:
Ok.

adschm commented on package/firmware/layerscape/ls-rcw/Makefile:
```diff
@@ -8,13 +8,13 @@
-PKG_VERSION:=LSDK-20.04
-PKG_RELEASE:=2
+PKG_VERSION:=LSDK-20.04-update-290520
+PKG_RELEASE:=2.1
```
PKG_RELEASE is the OpenWrt-internal version.
That means if you update PKG_VERSION, PKG_RELEASE goes back to "1" and is then incremented for "local" changes, until upstream is updated again ...

yangbolu1991 replied:
Thanks. Understand now.

adschm commented:
@yangbolu1991 

Can I pick only these four patches for now, or is there a dependency I'm missing:

```
layerscape: update kernel to LSDK-20.04-V5.4-update-290520
layerscape: update ls-rcw to LSDK-20.04-update-290520
layerscape: update u-boot to LSDK-20.04-update-290520
layerscape: update tfa to LSDK-20.04-update-290520
```

adschm commented:
@CHKDSK88 FYI.

adschm commented:
Consider submitting "perf: add libcap dependency" separately, either in a separate PR or via the mailing list. I fear it otherwise won't reach the correct people to review it.

adschm commented:
I've merged the sysupgrade PR and took the four update-290520 patches plus the NXP linux factory changes from this PR.

Note that I removed the lx2160a changes from ls-rcw patch, as that device is not merged yet. Take care to include the updated patchs in the lx2160a support patch when rebasing.

CHKDSK88 commented on package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-sdboot-uEnv.txt:
```diff
@@ -0,0 +1,9 @@
+fdtaddr=0x8f000000
+loadaddr=0x81000000
+fdt_high=0xffffffffffffffff
```
Please fix ```fdt_high``` value for proper work fit image.

yangbolu1991 replied:
Ok. Thanks.

CHKDSK88 replied:
Are You sure, that this board boot with `fdt_high=0xffffffffffffffff`?

CHKDSK88 commented on package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-sdboot-uEnv.txt:
```diff
@@ -0,0 +1,9 @@
+fdtaddr=0x8f000000
+loadaddr=0x81000000
+fdt_high=0xffffffffffffffff
```
Please fix fdt_high value for proper work fit image.

yangbolu1991 replied:
Ok, thanks.

CHKDSK88 replied:
My two boards (LS1012A-FRDM and LS1046A-RDB) won't start kernel from fit image with `fdt_high=0xffffffffffffffff`. Could You confirm, that it works for this boards?

CHKDSK88 reviewed (commented):
I will look again into this PR after rebase to master.

yangbolu1991 commented:
> 
> 
> Consider submitting "perf: add libcap dependency" separately, either in a separate PR or via the mailing list. I fear it otherwise won't reach the correct people to review it.

Thanks. Will send it separately.

yangbolu1991 commented:
> 
> 
> I've merged the sysupgrade PR and took the four update-290520 patches plus the NXP linux factory changes from this PR.
> 
> Note that I removed the lx2160a changes from ls-rcw patch, as that device is not merged yet. Take care to include the updated patchs in the lx2160a support patch when rebasing.

Thank you for cherry-picking. I will rebase the rest soon.

yangbolu1991 committed:
layerscape: some kmod packages enablement

Enable some kmod packages per board.

- Enable kmod-ahci-qoriq in default.
- Enable kmod-rtc-ds1307 and kmod-rtc-pcf85363 for LS1043ARDB.
- Enable kmod-rtc-pcf2127 for LS1046ARDB.

Signed-off-by: Biwen Li <biwen.li@nxp.com>
Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
layerscape: start MC firmware during u-boot booting

Split u-boot environment mc_init into mcinitcmd and dpl_apply.

- mcinitcmd will be run to start MC firmware during u-boot booting
  to make DPAA2 network available.

- dpl_apply will be run when booting kernel.

Signed-off-by: Biwen Li <biwen.li@nxp.com>
Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Pushed v2 patch-set.
Changes for v2:
- Rebased on latest master branch.
- Updated PKG_RELEASE for package internal changes.
- Dropped wildcards usage for board name in script.
- Dropped perf package patch and sent it separately.

Thanks!

CHKDSK88 commented on package/firmware/layerscape/ls-ddr-phy/Makefile:
```diff
@@ -0,0 +1,49 @@
+#
```
SDPX Header?

yangbolu1991 replied:
Let me add it.

CHKDSK88 replied:
Please move it to first line:

https://spdx.github.io/spdx-spec/appendix-V-using-SPDX-short-identifiers-in-source-files/#format-for-spdx-license-identifier

yangbolu1991 replied:
Ok.

yangbolu1991 commented:
@CHKDSK88 
For the fdt_high setting, during my previous tests, booting was fine.
But I just double checked, I indeed got booting issue. This happened when the dtb was loaded at address which was not 8-byte boundary.

So thanks for your reminding. We may have several ways to resolve, like setting proper fdt_high, or using bootm_size which may be better. And this may be also resolved by making kernel its specifying loadaddr for fdt (It seems current openwrt script does not support). Anyway, I will add the fix in new version patches.
Thanks.

dangowrt commented on package/firmware/layerscape/ls-ddr-phy/Makefile:
```diff
@@ -0,0 +1,49 @@
+PKG_SOURCE_URL:=https://github.com/NXP/ddr-phy-binary.git
+PKG_SOURCE_VERSION:=fbc036b88acb6c06ffed02c898cbae9856ec75ba
+PKG_MIRROR_HASH:=84e2bdea99384211971bb23ba9ed18b5839628ff2aa0738d2978bbdf841638cb
+PKG_BUILD_DEPENDS:=tfa-layerscape/host
```
PKG_LICENSE?

yangbolu1991 replied:
Already addressed since v3.

dangowrt reviewed (changes requested):
`PKG_LICENSE` is missing for all packages added. Please add that as well as `PKG_LICENSE_FILES` (if any).

CHKDSK88 commented:
@yangbolu1991 

> Anyway, I will add the fix in new version patches.

Thank You. I will wait for new version of PR.

yangbolu1991 commented:
Pushed v3 patch-set.
Changes for v3:
- Added SPDX header and PKG_LICENSE/PKG_LICENSE_FILES for new ls-ddr-phy package.
- Fixed SD booting issue by using bootm_size for new boards ls1046afrwy/ls1028ardb/lx2160ardb.

Thanks.

CHKDSK88 commented on target/linux/layerscape/base-files/etc/board.d/02_network:
```diff
@@ -14,6 +14,20 @@ case "$(board_name)" in
+			ip link set $enetc0 name eno0
+		fi
+		if [ -n $enetc2 ]; then
+			ip link set $enetc2 name eno2
```
IMO, this part should be in [05_layerscape_reorder_eth](https://github.com/openwrt/openwrt/blob/master/target/linux/layerscape/base-files/lib/preinit/05_layerscape_reorder_eth).

yangbolu1991 replied:
Ok.

CHKDSK88 commented on target/linux/layerscape/base-files/etc/board.d/02_network:
```diff
@@ -14,6 +14,20 @@ case "$(board_name)" in
+		fi
+		if [ -n $enetc2 ]; then
+			ip link set $enetc2 name eno2
+			ip link set eno2 up
```
Is it necessary? Other devices with DSA doesn't up interface manually.

yangbolu1991 replied:
It's necessary. This port is connecting to CPU port of the switch. It should be up always.

CHKDSK88 reviewed (commented):
@yangbolu1991 

Thank You for update. I mark some issues.

`bootm_size` works also with LS1012A-RDB and it's fine for me.

yangbolu1991 commented:
Pushed v4 patch-set.
Changes for v4:
- Moved SPDX to first line.
- Moved interface re-naming to 05_layerscape_reorder_eth.

yangbolu1991 commented:
Any comments on the new version patches?
Thanks.

CHKDSK88 commented on target/linux/layerscape/base-files/etc/board.d/02_network:
```diff
@@ -14,6 +14,12 @@ case "$(board_name)" in
 		;;
+	fsl,ls1028a-rdb | \
+	fsl,ls1028a-rdb-sdboot)
+		ip link set eno2 up
```
This script isn't launched at every boot.

It should be moved to preinit scripts. Examples:

https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/mt7621/base-files/lib/preinit/07_mt7621_bringup_dsa_master

https://github.com/openwrt/openwrt/blob/master/target/linux/mediatek/base-files/lib/preinit/05_set_preinit_iface

yangbolu1991 replied:
Done.

adschm commented:
Picked "layerscape: make restool depend on TARGET_layerscape_armv8_64b".

yangbolu1991 commented:
Pushed v5 patch-set.
Changes for v5:
- Moved "ip link set eno2 up" to preinit script.
- Rebased patches.

Thanks.

CHKDSK88 commented on package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-uEnv.txt:
```diff
@@ -0,0 +1,9 @@
+fdtaddr=0x8f000000
+loadaddr=0x81000000
+fdt_high=0xffffffffffffffff
```
Please add bootm_size to consolidate envs because https://github.com/openwrt/openwrt/commit/65f8d7360c041ed90724bcc97673b30ce704cc33 was merged.

yangbolu1991 replied:
Done.

CHKDSK88 commented on package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-uEnv.txt:
```diff
@@ -0,0 +1,9 @@
+fdtaddr=0x8f000000
+loadaddr=0x81000000
+fdt_high=0xffffffffffffffff
+initrd_high=0xffffffffffffffff
```
Please add bootm_size to consolidate envs because https://github.com/openwrt/openwrt/commit/65f8d7360c041ed90724bcc97673b30ce704cc33 was merged.

yangbolu1991 replied:
Done.

CHKDSK88 commented on package/boot/uboot-layerscape/files/fsl_lx2160a-rdb-uEnv.txt:
```diff
@@ -0,0 +1,11 @@
+fdtaddr=0x8f000000
+loadaddr=0x81000000
+fdt_high=0xffffffffffffffff
+initrd_high=0xffffffffffffffff
```
Please add bootm_size to consolidate envs because https://github.com/openwrt/openwrt/commit/65f8d7360c041ed90724bcc97673b30ce704cc33 was merged.

yangbolu1991 replied:
Done.

CHKDSK88 commented:
@yangbolu1991 

You need to rebase again. My PR was merged.

yangbolu1991 commented:
@CHKDSK88 
Good to see your pull request was merged. I rebased mine.

Changes for v6 patch-set,
- Rebased.
- Converted to use bootm_size env in u-boot for new added boards.

Thanks.

yangbolu1991 commented:
@adschm ,
Any comments on the new version?
Thanks.

yangbolu1991 commented:
@adschm ,
Will you merge the pull request?
Thanks.

adschm commented:
> @adschm ,
> Will you merge the pull request?
> Thanks.

Sorry, probably not anytime soon. This is too big for my currently available time, and layerscape is actually not one of my primary platforms/realms, so reviewing it doesn't come easy. (That's why I picked some of the straightforward patches beforehand BTW).

But there are other committer as well, and for a big PR like this I fear you will just have to wait a little longer.

I might have another quick look this weak, though, maybe I can at least pick parts of it again; but no promises ...

yangbolu1991 commented:
@adschm ,
Rebased and build test. No code changes.
I understand. I think I need to change my upstreaming strategy next time:)
These patches were from NXP released project and were tested on the code base.
I had thought it's easier for community acceptance after they were tested by us and sent together.
But I should have thought this would bring large code changes for reviewing one time. Sorry.

For this pull reqeust, the new boards support patches introduced large code changes. The other patches are minor changes.
I'm responsible to address any problem of conflict, build, and bug on your code base. Please feel free to review and merge.
6e0ceca layerscape: add LS1028ARDB board support
fabd01c layerscape: add FRWY-LS1046A board support
c07448c layerscape: add LX2160ARDB (Rev2.0 silicon) board support
b55e321 layerscape: add ddr-phy package

Thanks.

adschm commented on package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-uEnv.txt:
```diff
@@ -0,0 +1,8 @@
+bootm_size=0x10000000
+hwconfig=fsl_ddr:bank_intlv=auto
+qspi_boot=sf probe 0:0;sf read $fdtaddr f00000 100000;sf read $loadaddr 1000000 1000000;bootm $loadaddr - $fdtaddr
+bootargs=root=/dev/mtdblock9 rootfstype=squashfs,jffs2 noinitrd earlycon=uart8250,mmio,0x21c0500 console=ttyS0,115200 mtdparts=1550000.spi:1m(bl2),4m(fip),1m(u-boot-env),3m(reserved-1),256k(fman),5888k(reserved-2),1m(dtb),16m(kernel),32m(rootfs),48m@0x1000000(firmware)
```
Shouldn't that be 49m@0xf00000 due to dtb inclusion in image?
@yangbolu1991 @CHKDSK88 
Please just answer here and do not update/repush, as I've already picked stuff.

CHKDSK88 replied:
It should be `49m@0xf00000`. Now sysupgrade for this devices are broken.

adschm commented on package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-uEnv.txt:
```diff
@@ -0,0 +1,8 @@
+bootm_size=0x10000000
+hwconfig=fsl_ddr:bank_intlv=auto
+xspi_boot=sf probe 0:0;sf read $fdtaddr f00000 100000;sf read $loadaddr 1000000 1000000;bootm $loadaddr - $fdtaddr
+bootargs=root=/dev/mtdblock8 rootfstype=squashfs,jffs2 noinitrd earlycon=uart8250,mmio,0x21c0500 console=ttyS0,115200 mtdparts=20c0000.spi:1m(bl2),4m(fip),1m(u-boot-env),3m(reserved-1),256k(fman),5888k(reserved-2),1m(dtb),16m(kernel),32m(rootfs),48m@0x1000000(firmware)
```
same question as above: 49m@0xf00000 ?

adschm commented on target/linux/layerscape/base-files/etc/board.d/02_network:
```diff
@@ -14,6 +14,11 @@ case "$(board_name)" in
+	fsl,ls1028a-rdb-sdboot)
+		ucidef_set_interface_lan "swp0 swp1 swp2 swp3"
+		ucidef_set_interface_wan "eno0"
+		;;
```
This should be moved for proper sorting.

adschm commented on target/linux/layerscape/base-files/lib/preinit/05_layerscape_reorder_eth:
```diff
@@ -17,6 +17,19 @@ reorder_layerscape_interfaces() {
 		ip link set fm1-mac4 name eth5
 		;;
+	fsl,ls1028a-rdb | \
+	fsl,ls1028a-rdb-sdboot)
```
Sorting.

adschm commented on package/boot/uboot-layerscape/files/fsl_lx2160a-rdb-uEnv.txt:
```diff
@@ -0,0 +1,9 @@
+hwconfig=fsl_ddr:bank_intlv=auto
+mc_init=sf probe 0:0;sf read 80000000 a00000 300000;sf read 80300000 e00000 100000;fsl_mc start mc 80000000 80300000;sf read 80400000 d00000 100000;fsl_mc apply dpl 80400000
+xspi_boot=sf probe 0:0;sf read $fdtaddr f00000 100000;sf read $loadaddr 1000000 1000000;bootm $loadaddr - $fdtaddr
+bootargs=root=/dev/mtdblock9 rootfstype=squashfs,jffs2 noinitrd earlycon=pl011,mmio32,0x21c0000 console=ttyAMA0,115200 mtdparts=20c0000.spi-0:1m(bl2),4m(fip),1m(u-boot-env),4m(reserved-1),3m(mc),1m(dpl),1m(dpc),1m(dtb),16m(kernel),32m(rootfs),48m@0x1000000(firmware)
```
49m@0xf00000 ?

adschm commented:
This would be the next subset of patches I'd take:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/layers

However, this still needs confirmation for the 48m/49m question, which is _not_ included yet.
I haven't build-tested yet, so maybe I'll update if there are problems.

yangbolu1991 commented:
@adschm @CHKDSK88 ,
Thank you very much for pointing out the firmware partition issue. Yes. It should be 49m@0xf00000.
And the sorting problem in scripts needs fix too.

Should I wait for cherry-picked patches merged by mainline, and rebase this pull request with fix?
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/layers

Or update the pull request now?
Thank you.

adschm commented:
@yangbolu1991 If you think my patches are okay, I'd merge them first (just with the 49m changed), and then you could rebase your remaining stuff afterwards.

yangbolu1991 commented:
@adschm All your patches look good. I will wait for merging them, and rebase the remaining soon after that.
Thank you very much.

adschm commented:
Patches merged.

yangbolu1991 commented:
Rebased the rest patches, with sorting problem and firmware address problem addressed.
Thanks a lot.

CodeFetch commented:
@yangbolu1991 You have conflicting files. Why do you disable so many kernel symbols? Aren't they needed for something else?
